### ISSUE TITLE
Allow running setup hooks with multithreading disabled

### ISSUE THREAD

**Issue by andy0130tw** [haskell/cabal#9563]
As of Cabal v3.10, when building a package that specifies custom setup hooks, i.e. `Setup.hs`, Cabal will append `-threaded` to the list of extra options, as seen in:

https://github.com/haskell/cabal/blob/3.10/cabal-install/src/Distribution/Client/SetupWrapper.hs#L808-L811

GHCJS is an exception, it opts for `-build-runner` and does not specify `-threaded`. Unfortunately, if the backend is WASM, passing `-threaded` will result in failure due to WebAssembly's lack of support for threading. As of now, [the WASM backend is in tech preview](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/wasm.html), I suggest adding an option to explicitly disable threading, or making it a special case for WASM as in the JS backend.

**Comment by Mikolaj** [comment]
Well spotted. That sounds reasonable. However, this is a 9 years old snippet of code and I'm clueless about the big picture. @hsyl20, is this still how `GHCJS` should be treated? BTW, does it affect the old GHCJS or the new JS cross-compilation in GHC or both? If this does not affect the GHC cross-compiling to JS, why should it affect GHC cross-compiling to WebAssembly?

**Comment by hsyl20** [comment]
I don't know much about GHCJS. I would probably leave the code untouched until it is removed.

`Setup.hs` built with the JS backend should work since 9.8 (as long as there is no call to C code). So nothing special required compared to native backends.

The JS backend has its own RTS and just ignores the `-threaded` flag iirc. The Wasm backend reuses the C RTS and may not ignore the flag. It explains the issue.

**Comment by andy0130tw** [comment]
Yes, when compiling the WASM backend for GHC, the C RTS is compiled. Further digging into the GHC codebase reveals that, in all architectures other than Wasm32 and JavaScript backend, both threaded and non-threaded RTS are compiled. This is controlled by [`targetSupportsThreadedRts`](https://gitlab.haskell.org/ghc/ghc/-/blob/master/hadrian/src/Oracles/Flag.hs#L97) in Hadrian:

```hs
-- | Does the target support threaded RTS?
targetSupportsThreadedRts :: Action Bool
targetSupportsThreadedRts = do
    bad_arch <- anyTargetArch [ ArchWasm32, ArchJavaScript ]
    return $ not bad_arch
```

And it [determines](https://gitlab.haskell.org/ghc/ghc/-/blob/master/hadrian/src/Settings/Default.hs#L206) whether threaded ways of RTS are built (excerpted):

```hs
-- | Default build ways for the RTS.
defaultRtsWays :: Ways
defaultRtsWays = Set.fromList <$>
  mconcat
  [ ...
  , notStage0 ? targetSupportsThreadedRts ? pure [threaded, threadedProfiling, threadedDebugProfiling, threadedDebug]
  , ...
  ]
```

We know that `-threaded` selects the version of RTS to link, and that it is not supported by either WASM or JS backend (not knowing if `ArchJavaScript` also applies to GHCJS). Hence it can be safely used in all other architectures. I think we only need to deal with the WASM case.

**Comment by Mikolaj** [comment]
Thank you. Let's do it!
